### ISSUE TITLE
Adding tests that use Smocks interferes with branch points 

### ISSUE THREAD

**Issue by aburgett87** [OpenCover/opencover#364]
Using Smocks (https://github.com/vanderkleij/Smocks) seems to distort the calculation of branch points. If I add one test that uses Smocks all branch points in all tests (except for the test using Smock) increase by 2, adding 2 tests increase the branch points by a further 2, as far as I can see branch_points = branch_points +  2 * amount_smock_tests.

Any ideas on why this may happen?

**Comment by ddur** [comment]
With ILSpy, you can try to see what is going on.

**Comment by bogan87** [comment]
On the assembly under test there isn't any difference in the IL if Smocks is used or not. It appears to be something with the tests themselves. The branch points go back to normal if I ignore the test fixture containing Smocks.

I should also report that I'm using, Moq (4.2.1507.118), NUnint (2.6.4.14350) and Smocks (0.5.22.0)

**Comment by ddur** [comment]
I guess any Mock or Smock will create additional branches,
Maybe you can you filter it out?

**Comment by bogan87** [comment]
I thought so as well, but it effects every test, not just the test using Smock.

e.g. If I have a test suite including 4 smocks tests, the following will generate 10 branch points,
```c#
//Code under test example
if(tempString.Equals("Hello")) //Ten branch points here instead of two
{
   DoSomething()
}
```

When you say filter it out, do you mean the branches created or the tests themselves?

**Comment by ddur** [comment]
If assembly under test is unchanged, then new branches are created somewhere else.
Usually I do filter out test suite coverage, what I want to know is only coverage of system/assembly under test. Do you really need coverage of your test?

**Comment by bogan87** [comment]
I've just done a bit of research into Smocks and it turns out that it rewrites and then loads the new assemblies to use the faked modules. This seems to add to the list of modules that NUnit exposes and OpenCover picks up and processes it, because it has the same ModuleName as the assembly under test. I probably could find a way of filtering out the re-written assemblies, they all seem to go to the system's temp directory.

I don't really need to cover those tests, I was just trying to be a perfectionist and get 100% coverage haha. It really was nothing more than an experiment.

Would adding a new switch to the command line to skip modules with same name be something of interest? I'm more than happy to do it.

**Comment by ddur** [comment]
Why not? If it passes tests and doesn't break anything else  ....
Did you try to select only namespaces of your assembly under test?

**Comment by ddur** [comment]
If I understood you correctly, both original and by-smocks-rewritten assemblies are loaded into memory?

**Comment by ddur** [comment]
@bogan87 
If I understood you correctly, both original and by-smocks-rewritten assemblies are loaded into memory?

**Comment by bogan87** [comment]
Yep, that's exactly it.

**Comment by ddur** [comment]
@bogan87 
Good, you can use one of [installers from latest build](https://ci.appveyor.com/project/sawilde/opencover/build/4.6.324/artifacts) and try exclusion filter like ```-<path to smock assemblies>[*]*```

In example ```-<C:\project\smocks\*>[*]*```

Let us know if that helps

**Comment by ddur** [comment]
If profiler checks filter for process that is only .exe, full path process filter will not help here.
We need to add full path assembly filter..

**Comment by sawilde** [comment]
https://www.nuget.org/packages/OpenCover/4.6.515-rc

**Comment by sawilde** [comment]
The excluded paths feature

We have published the release at https://www.nuget.org/packages/OpenCover/4.6.519 and https://github.com/OpenCover/opencover/releases/tag/4.6.519

**Comment by unittester57** [comment]
Hi all, I am new to Smocks and OpenCover. Right now this issue is bothering me as I wont get 100% brnch coverage. Can someone help me in providing example syntax how to exclude Smocks assemblies for code coverage. Below is what I have tried.

`"C:\Users\uname\Documents\Visual Studio 2015\Projects\firstproject\packages\OpenCover.4.6.519\tools\OpenCover.Console.exe" -register:Path32  "-target:MSTest.exe" -targetargs:"/testcontainer:firstprojectTests1.dll" -excludedirs:<"C:\Users\uname\Documents\Visual Studio 2015\Projects\firstproject\firstprojectTests1\bin\Debug\Smocks.dll">[]` 

I have pointed to Smocks.dll as I did-not see where Smocks assemblies are created. Any help is appreciated.
